purchases/src/main/kotlin/com/revenuecat/purchases/storage/FileRepository.kt

```diff
@@ -0,0 +1,197 @@
+ */
+internal interface FileRepositoryType {
+    /**
+     * Prefetch files at the given urls.
+     * @param urls An array of URL to fetch data from.
+     * @return The Job that is executing the prefetch
+     */
+    fun prefetch(urls: List<URL>)
+
+    /**
+     * Create and/or get the cached file url.
+     * @param url The url for the remote data to cache into a file.
+     * @return The local file's URI where the data can be found after caching is complete.
+     */
+    suspend fun generateOrGetCachedFileURL(url: URL): URI
+}
+
+/**
+ * The file manager is a service capable of storing data and returning the URL where that stored data exists.
+ */
+internal interface LargeItemCacheType {
+    fun generateLocalFilesystemURI(remoteURL: URL): URI?
+    fun cachedContentExists(uri: URI): Boolean
+    fun saveData(data: InputStream, uri: URI)
+}
+
+internal class FileRepository(
+    @get:VisibleForTesting
+    internal val store: KeyedDeferredValueStore<URL, URI> = KeyedDeferredValueStore<URL, URI>(),
+    private val fileCacheManager: LargeItemCacheType,
+    private val scope: CoroutineScope = CoroutineScope(Dispatchers.IO + NonCancellable),
+    private val logHandler: LogHandler = currentLogHandler,
+    private val urlConnectionFactory: UrlConnectionFactory = DefaultUrlConnectionFactory(),
+) : FileRepositoryType {
+
+    // Convenience constructor for Android
+    constructor(
+        context: Context,
+    ) : this(
+        fileCacheManager = FileCache(context),
+    )
+
+    override fun prefetch(urls: List<URL>) {
+        scope.launch {
+            urls.forEach { url ->
+                try {
+                    generateOrGetCachedFileURL(url)
+                } catch (e: IOException) {
+                    logHandler.e("FileRepository", "Prefetch failed for $url: $e", e)
+                }
+            }
+        }
+    }
+
+    override suspend fun generateOrGetCachedFileURL(url: URL): URI {
+        return store.getOrPut(url) {
+            scope.async {
+                val cachedUri = fileCacheManager.generateLocalFilesystemURI(remoteURL = url)
+                    ?: {
+                        val error = Error.FailedToCreateCacheDirectory(url.toString())
+                        logHandler.e("FileRepository", "Failed to create cache directory for $url", error)
+                        throw error
+                    }()
+
+                if (fileCacheManager.cachedContentExists(cachedUri)) {
+                    return@async cachedUri
+                }
+
+                val data = downloadFile(url = url)
+                saveCachedFile(cachedUri, data)
+                return@async cachedUri
+            }
+        }.await()
+    }
+
+    private suspend fun downloadFile(url: URL): InputStream {
+        return try {
+            withContext(Dispatchers.IO) {
+                verboseLog { "Downloading remote font from $url" }
+
+                val connection = urlConnectionFactory.createConnection(url.toString())
```

Comment:
Hmm with this we're not disconnecting the connection anymore right? I think we need to refactor this so we disconnect once the download has finished. Either joining the download + saving to a file, so we disconnect once everything has been downloaded, or just save to a temp file when downloading, then later just move the file to the final location.

Reply:
@tonidero I changed things up a bit, would you give that a look?